### ISSUE TITLE
Ollama Not Respecting Structured Outputs

### ISSUE THREAD

**Issue by mcmah309** [ollama/ollama#8063]
### What is the issue?

Example:
```sh
curl -X POST http://localhost:11434/api/chat -H "Content-Type: application/json" -d '{
  "model": "llama3.2",
  "messages": [{"role": "user", "content": "Create a random new user"}],
  "stream": false,
  "format": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "array",
    "minItems": 1,
    "maxItems": 1,
    "uniqueItems": true,
    "items": {
      "oneOf": [
        {
          "type": "object",
          "properties": {
            "function_name": {
              "const": "calculate_total"
            },
            "items": {
              "type": "array",
              "items": {
                "type": "object",
                "properties": {
                  "name": {"type": "string"},
                  "price": {"type": "number", "minimum": 0}
                },
                "required": ["name", "price"]
              }
            },
            "tax_rate": {"type": "number", "minimum": 0, "maximum": 1}
          },
          "required": ["function_name", "items"]
        },
        {
          "type": "object",
          "properties": {
            "function_name": {
              "const": "send_email"
            },
            "to": {
              "type": "string",
              "format": "email"
            },
            "subject": {"type": "string"},
            "body": {"type": "string"},
            "attachments": {
              "type": "array",
              "items": {
                "type": "object",
                "properties": {
                  "filename": {"type": "string"},
                  "content": {"type": "string"}
                },
                "required": ["filename", "content"]
              }
            }
          },
          "required": ["function_name", "to", "subject", "body"]
        },
        {
          "type": "object",
          "properties": {
            "function_name": {
              "const": "create_user"
            },
            "username": {"type": "string"},
            "email": {
              "type": "string", 
              "format": "email"
            },
            "password": {"type": "string", "minLength": 8},
            "role": {
              "type": "string",
              "enum": ["admin", "user", "editor"]
            }
          },
          "required": ["function_name", "username", "email", "password"]
        }
      ]
    }
  }
}'
```
output:
```json
{
    "model": "llama3.2",
    "created_at": "2024-12-12T06:12:38.576392006Z",
    "message": {
        "role": "assistant",
        "content": "[]"
    },
    "done_reason": "stop",
    "done": true,
    "total_duration": 107634991,
    "load_duration": 26643655,
    "prompt_eval_count": 30,
    "prompt_eval_duration": 6000000,
    "eval_count": 2,
    "eval_duration": 73000000
}
```
content part of the output should be something like
```json
[
{
    "function_name": "create_user",
    "username": "david123",
    "email": "johndoe@gmail.com",
   "password": "kasdsd!dkDj",
}
]
```

### OS

Linux

### GPU

Nvidia

### CPU

AMD

### Ollama version

5.0.0

**Comment by mcmah309** [comment]
This issue does not appear to be related to llama.cpp as it works with the lastest build
```console
./build/bin/llama-cli -m ./models/7B/ggml-model-f16.gguf --prompt "Create a new user" --json-schema '{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "array",
    "minItems": 1,
    "maxItems": 1,
    "uniqueItems": true,
    "items": {
        "oneOf": [
            {
                "type": "object",
                "properties": {
                    "function_name": {
                        "const": "calculate_total"
                    },
                    "items": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "name": {
                                    "type": "string"
                                },
                                "price": {
                                    "type": "number",
                                    "minimum": 0
                                }
                            },
                            "required": [
                                "name",
                                "price"
                            ]
                        }
                    },
                    "tax_rate": {
                        "type": "number",
                        "minimum": 0,
                        "maximum": 1
                    }
                },
                "required": [
                    "function_name",
                    "items"
                ]
            },
            {
                "type": "object",
                "properties": {
                    "function_name": {
                        "const": "send_email"
                    },
                    "to": {
                        "type": "string",
                        "format": "email"
                    },
                    "subject": {
                        "type": "string"
                    },
                    "body": {
                        "type": "string"
                    },
                    "attachments": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "filename": {
                                    "type": "string"
                                },
                                "content": {
                                    "type": "string"
                                }
                            },
                            "required": [
                                "filename",
                                "content"
                            ]
                        }
                    }
                },
                "required": [
                    "function_name",
                    "to",
                    "subject",
                    "body"
                ]
            },
            {
                "type": "object",
                "properties": {
                    "function_name": {
                        "const": "create_user"
                    },
                    "username": {
                        "type": "string"
                    },
                    "email": {
                        "type": "string",
                        "format": "email"
                    },
                    "password": {
                        "type": "string",
                        "minLength": 8
                    },
                    "role": {
                        "type": "string",
                        "enum": [
                            "admin",
                            "user",
                            "editor"
                        ]
                    }
                },
                "required": [
                    "function_name",
                    "username",
                    "email",
                    "password"
                ]
            }
        ]
    }
}'
build: 4313 (4064c0e3) with gcc (GCC) 11.5.0 for x86_64-unknown-linux-gnu
main: llama backend init
main: load the model and apply lora adapter, if any
llama_model_loader: loaded meta data with 19 key-value pairs and 291 tensors from ./models/7B/ggml-model-f16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = checkpoints
llama_model_loader: - kv   2:                       llama.context_length u32              = 2048
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 11008
llama_model_loader: - kv   6:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv   7:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   8:              llama.attention.head_count_kv u32              = 32
llama_model_loader: - kv   9:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                       llama.rope.freq_base f32              = 10000.000000
llama_model_loader: - kv  11:                          general.file_type u32              = 1
llama_model_loader: - kv  12:                       tokenizer.ggml.model str              = llama
llama_model_loader: - kv  13:                      tokenizer.ggml.tokens arr[str,32000]   = ["<unk>", "<s>", "</s>", "<0x00>", "<...
llama_model_loader: - kv  14:                      tokenizer.ggml.scores arr[f32,32000]   = [0.000000, 0.000000, 0.000000, 0.0000...
llama_model_loader: - kv  15:                  tokenizer.ggml.token_type arr[i32,32000]   = [2, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...
llama_model_loader: - kv  16:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  17:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  18:            tokenizer.ggml.padding_token_id u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:  226 tensors
llm_load_vocab: special_eos_id is not in special_eog_ids - the tokenizer config may be incorrect
llm_load_vocab: special tokens cache size = 3
llm_load_vocab: token to piece cache size = 0.1684 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = SPM
llm_load_print_meta: n_vocab          = 32000
llm_load_print_meta: n_merges         = 0
llm_load_print_meta: vocab_only       = 0
llm_load_print_meta: n_ctx_train      = 2048
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_swa            = 0
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 1
llm_load_print_meta: n_embd_k_gqa     = 4096
llm_load_print_meta: n_embd_v_gqa     = 4096
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 11008
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 10000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 2048
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: ssm_dt_b_c_rms   = 0
llm_load_print_meta: model type       = 7B
llm_load_print_meta: model ftype      = F16
llm_load_print_meta: model params     = 6.74 B
llm_load_print_meta: model size       = 12.55 GiB (16.00 BPW) 
llm_load_print_meta: general.name     = checkpoints
llm_load_print_meta: BOS token        = 1 '<s>'
llm_load_print_meta: EOS token        = 2 '</s>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: PAD token        = 2 '</s>'
llm_load_print_meta: LF token         = 13 '<0x0A>'
llm_load_print_meta: EOG token        = 2 '</s>'
llm_load_print_meta: max token length = 48
llm_load_tensors:   CPU_Mapped model buffer size = 12853.02 MiB
...................................................................................................
llama_new_context_with_model: n_seq_max     = 1
llama_new_context_with_model: n_ctx         = 4096
llama_new_context_with_model: n_ctx_per_seq = 4096
llama_new_context_with_model: n_batch       = 2048
llama_new_context_with_model: n_ubatch      = 512
llama_new_context_with_model: flash_attn    = 0
llama_new_context_with_model: freq_base     = 10000.0
llama_new_context_with_model: freq_scale    = 1
llama_new_context_with_model: n_ctx_pre_seq (4096) > n_ctx_train (2048) -- possible training context overflow
llama_kv_cache_init:        CPU KV buffer size =  2048.00 MiB
llama_new_context_with_model: KV self size  = 2048.00 MiB, K (f16): 1024.00 MiB, V (f16): 1024.00 MiB
llama_new_context_with_model:        CPU  output buffer size =     0.12 MiB
llama_new_context_with_model:        CPU compute buffer size =   296.01 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 1
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
main: llama threadpool init, n_threads = 12
main: model was trained on only 2048 context tokens (4096 specified)

system_info: n_threads = 12 (n_threads_batch = 12) / 24 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

sampler seed: 856076181
sampler params: 
	repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
	dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = -1
	top_k = 40, top_p = 0.950, min_p = 0.050, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000, temp = 0.800
	mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampler chain: logits -> logit-bias -> penalties -> dry -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist 
generate: n_ctx = 4096, n_batch = 2048, n_predict = -1, n_keep = 1

 Create a new user[
              {
                "function_name": "create_user",
                "username": "username",
                "email": "email",
                "password": "password",
                "role": "admin"
              }
            ]
         											[end of text]


llama_perf_sampler_print:    sampling time =       3.27 ms /    78 runs   (    0.04 ms per token, 23853.21 tokens per second)
llama_perf_context_print:        load time =    2558.15 ms
llama_perf_context_print: prompt eval time =     660.60 ms /     5 tokens (  132.12 ms per token,     7.57 tokens per second)
llama_perf_context_print:        eval time =   20656.87 ms /    72 runs   (  286.90 ms per token,     3.49 tokens per second)
llama_perf_context_print:       total time =   21379.94 ms /    77 tokens
```

**Comment by rick-github** [comment]
What's the response like if you import ./models/7B/ggml-model-f16.gguf into ollama?

**Comment by mcmah309** [comment]
modelfile

```
FROM /home/henry/work/misc/llama.cpp/models/7B/ggml-model-f16.gguf
```
Create
```console
ollama create ggml-model-f16 -f ~/work/misc/llama.cpp/models/7B/ggml-model-f16.gguf
```
Test
```console
curl -X POST http://localhost:11434/api/chat -H "Content-Type: application/json" -d '{
  "model": "ggml-model-f16",
  "messages": [{"role": "user", "content": "Create a random new user"}],
  "stream": false,
  "format": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "array",
    "minItems": 1,
    "maxItems": 1,
    "uniqueItems": true,
    "items": {
      "oneOf": [
        {
          "type": "object",
          "properties": {
            "function_name": {
              "const": "calculate_total"
            },
            "items": {
              "type": "array",
              "items": {
                "type": "object",
                "properties": {
                  "name": {"type": "string"},
                  "price": {"type": "number", "minimum": 0}
                },
                "required": ["name", "price"]
              }
            },
            "tax_rate": {"type": "number", "minimum": 0, "maximum": 1}
          },
          "required": ["function_name", "items"]
        },
        {
          "type": "object",
          "properties": {
            "function_name": {
              "const": "send_email"
            },
            "to": {
              "type": "string",
              "format": "email"
            },
            "subject": {"type": "string"},
            "body": {"type": "string"},
            "attachments": {
              "type": "array",
              "items": {
                "type": "object",
                "properties": {
                  "filename": {"type": "string"},
                  "content": {"type": "string"}
                },
                "required": ["filename", "content"]
              }
            }
          },
          "required": ["function_name", "to", "subject", "body"]
        },
        {
          "type": "object",
          "properties": {
            "function_name": {
              "const": "create_user"
            },
            "username": {"type": "string"},
            "email": {
              "type": "string", 
              "format": "email"
            },
            "password": {"type": "string", "minLength": 8},
            "role": {
              "type": "string",
              "enum": ["admin", "user", "editor"]
            }
          },
          "required": ["function_name", "username", "email", "password"]
        }
      ]
    }
  }
}'
{"model":"ggml-model-f16","created_at":"2024-12-13T00:50:01.977449384Z","message":{"role":"assistant","content":"[1]\n   "},"done_reason":"stop","done":true,"total_duration":4782538698,"load_duration":4503104076,"prompt_eval_count":6,"prompt_eval_duration":140000000,"eval_count":6,"eval_duration":137000000}
```
Looks like the same issue. So it is not dependent on the model here.

**Comment by ParthSareen** [comment]
Hey! Thanks for reporting. We're currently just using the grammar generation from llama.cpp so I am a bit surprised. Will check this out and report back 🫡

**Comment by mcmah309** [comment]
It shouldn’t depend on the model based on how I understand structured outputs are implemented. It works on llama.cpp, which ollama is just a wrapper around, so it should work.

**Comment by devlux76** [comment]
If the model isn't code centric it stands to reason it would get confused on a large json schema. Unsure why you're seeing a difference between llama.cpp and ollama. However, my understanding is that ollama is no longer strictly speaking a wrapper. I believe they have their own internals.

**Comment by mcmah309** [comment]
In layman's terms, structure outputs is implemented by the selecting the most probable token that is still valid for the schema, based on the models token probability distribution. It shouldn't depend on the model used. Ollama uses a fork of llama.cpp at the core. So my guess is some fixes were never merged into their fork.

**Comment by andygill** [comment]
Here is a python program that can be used to generate a working, and failing, examples.
I assume it is the same root bug, and the schemas are smaller that above.

Changing the `schema = ...` line to use `schema_works`, which is a slightly different schema, works.
The two schemas should parse the same JSON - the only difference is the  ordering in "$defs".

I have examples of pydantic-generated schemas which fail because of this bug.

```
from ollama import chat
import json
from jsonschema import validate, ValidationError

#
# Schema originally generated by pydantic. Depending
# on the alphabetical order of the names of the classes,
# both failing and working schemas can be generated.
# This is the smallest example I could find, and
# I'm using client version is 0.5.5.
#
# The *only* difference is the order of "C" and "A" in "$defs",
# This should not make any difference: both are supposed to be the
# same schema.
schema_fails = {
    "$defs": {
        "C": {
            "properties": {"payload": {"$ref": "#/$defs/A"}},
            "required": ["payload"],
            "title": "C",
            "type": "object",
        },
        "A": {
            "properties": {"payload": {"$ref": "#/$defs/E"}},
            "required": ["payload"],
            "title": "A",
            "type": "object",
        },
        "E": {"type": "boolean"},
    },
    "properties": {"payload": {"$ref": "#/$defs/C"}},
    "required": ["payload"],
    "title": "B",
    "type": "object",
}

schema_works = {
    "$defs": {
        "A": {
            "properties": {"payload": {"$ref": "#/$defs/E"}},
            "required": ["payload"],
            "title": "A",
            "type": "object",
        },
        "C": {
            "properties": {"payload": {"$ref": "#/$defs/A"}},
            "required": ["payload"],
            "title": "C",
            "type": "object",
        },
        "E": {"type": "boolean"},
    },
    "properties": {"payload": {"$ref": "#/$defs/C"}},
    "required": ["payload"],
    "title": "B",
    "type": "object",
}

schema = schema_fails  # or schema_works

response = chat(
    model="mistral-nemo:latest",
    messages=[
        {
            "role": "user",
            "content": "I want to choose true, inside A, inside C, inside B",
        }
    ],
    format=schema,
    options={"temperature": 0},  # Make responses more deterministic
)
valid_json = json.loads(response.message.content)
print(json.dumps(valid_json))
validate(instance=valid_json, schema=schema)
print("validates!")
```

**Comment by ericmjl** [comment]
I'm seeing structured generation working fine on my Mac, but not on Linux. Perhaps there might be a clue here?

**Comment by wanderr** [comment]
I seem to be having the same problem. In my case it works fine until context gets large

**Comment by ParthSareen** [comment]
Digging a bit deeper into where the issue is. @mcmah309 are you comparing quantized vs. unquantized? We default to `q4_k_m` which could possibly be hiding the bug in llama.cpp. I have a couple gut checks where I think it could be happening but knowing the above would definitely help :)

@ericmjl are you still seeing the variance in generation?

@wanderr does bumping up the ctx size just straightaway fix it?

**Comment by joshua-lehmann** [comment]
I can reproduce the same class of issue with a minimal standalone Python script using `/api/chat` and `format=<json schema>`. No external app/framework is involved.

This seems related to this issue rather than only the previously fixed `think=false` Gemma4 bug. I saw that #15678 fixed the specific case where `format` was ignored with `think=false` for Gemma4, but the repro below still returns schema-invalid values with both omitted `think` and `think=False`.

@ParthSareen @nicoloboschi in case this helps with triage.

Environment:
- OS: Windows 11
- Ollama version: `0.30.5`
- Python version: `3.13.14`
- Python `ollama` package version: unknown
- Models:
  - `gemma4:e2b-it-qat`
  - `gemma4:e4b-it-qat`
  - `gemma4:12b-it-qat`
- Tested both with omitted `think` and `think=False`.

`pip show ollama` output:
    Name: ollama
    Version: 0.6.2
    Summary: The official Python client for Ollama.
    Requires: httpx, pydantic

Minimal repro:

    import json
    import sys
    from typing import Any

    from ollama import Client

    try:
        from jsonschema import Draft202012Validator
    except ImportError:
        print("Missing dependency: pip install jsonschema ollama")
        sys.exit(1)


    MODELS = [
        "gemma4:e2b-it-qat",
        "gemma4:e4b-it-qat",
        "gemma4:12b-it-qat",
    ]

    SCHEMA: dict[str, Any] = {
        "type": "object",
        "properties": {
            "document_type": {
                "type": "string",
                "enum": ["Rechnung", "Beleg", "Medizinisches Dokument"],
            },
            "tags": {
                "type": "array",
                "items": {
                    "type": "string",
                    "enum": ["Gesundheit", "Finanzen", "Auto"],
                },
                "uniqueItems": True,
            },
            "created": {
                "type": "string",
                "pattern": r"^\d{4}-\d{2}-\d{2}$",
            },
        },
        "required": ["document_type", "tags", "created"],
        "additionalProperties": False,
    }

    PROMPT = """
    Return JSON only matching the provided schema.

    Use these values:
    - document_type: Finanzen
    - tags: Medizinisches Dokument, Zertifikat, Health
    - created: 2023-10-27T10:00:00Z

    Also include extra fields: author, status.
    """


    def strip_markdown_fence(text: str) -> str:
        text = text.strip()
        if text.startswith("```"):
            lines = text.splitlines()
            if lines and lines[0].startswith("```"):
                lines = lines[1:]
            if lines and lines[-1].strip() == "```":
                lines = lines[:-1]
            return "\n".join(lines).strip()
        return text


    def validate_content(content: str) -> list[str]:
        cleaned = strip_markdown_fence(content)

        try:
            parsed = json.loads(cleaned)
        except json.JSONDecodeError as exc:
            return [f"invalid JSON: {exc}"]

        validator = Draft202012Validator(SCHEMA)
        errors = sorted(validator.iter_errors(parsed), key=lambda e: list(e.path))

        return [
            f"{'/'.join(map(str, error.path)) or '<root>'}: {error.message}"
            for error in errors
        ]


    def run_case(client: Client, model: str, think_mode: str) -> None:
        kwargs: dict[str, Any] = {
            "model": model,
            "messages": [{"role": "user", "content": PROMPT}],
            "format": SCHEMA,
            "options": {
                "temperature": 0,
                "num_ctx": 8192,
                "num_predict": 512,
            },
            "stream": False,
        }

        if think_mode == "false":
            kwargs["think"] = False

        response = client.chat(**kwargs)
        content = response.message.content

        print("=" * 80)
        print(f"model={model} think={think_mode}")
        print("- raw content:")
        print(content)

        errors = validate_content(content)
        print("- local jsonschema validation:")
        if errors:
            print("FAILED")
            for error in errors:
                print(f"  - {error}")
        else:
            print("PASSED")


    def main() -> None:
        client = Client(host="http://localhost:11434")

        for think_mode in ["omit", "false"]:
            for model in MODELS:
                run_case(client, model, think_mode)


    if __name__ == "__main__":
        main()

Observed:
All tested cases failed local JSON Schema validation even though `format=SCHEMA` was passed.

Example from `gemma4:e4b-it-qat` with omitted `think`:

    {
      "document_type": "Finanzen",
      "tags": [
        "Medizinisches Dokument",
        "Zertifikat",
        "Health"
      ],
      "created": "2023-10-27T10:00:00Z",
      "author": null,
      "status": null
    }

Validation errors:

    FAILED
      - <root>: Additional properties are not allowed ('author', 'status' were unexpected)
      - created: '2023-10-27T10:00:00Z' does not match '^\\d{4}-\\d{2}-\\d{2}$'
      - document_type: 'Finanzen' is not one of ['Rechnung', 'Beleg', 'Medizinisches Dokument']
      - tags/0: 'Medizinisches Dokument' is not one of ['Gesundheit', 'Finanzen', 'Auto']
      - tags/1: 'Zertifikat' is not one of ['Gesundheit', 'Finanzen', 'Auto']
      - tags/2: 'Health' is not one of ['Gesundheit', 'Finanzen', 'Auto']

Also reproduced with `think=False`. Example from `gemma4:12b-it-qat`:

    {
      "document_type": "Finanzen",
      "tags": [
        "Medizinisches Dokument",
        "Zertifikat",
        "Health"
      ],
      "created": "2023-10-27T10:00:00Z",
      "author": null,
      "status": null
    }

Expected:
The returned `message.content` should satisfy the provided JSON Schema, especially `enum`, `required`, `pattern`, and `additionalProperties: false`, or the request should fail clearly instead of returning HTTP 200 / `done_reason=stop` with schema-invalid content.

**Comment by rick-github** [comment]
The fact that you have to strip out a `json` header and that you can ask the model to add tags that aren't in the schema is an indication that the schema is not being applied.  If you look in the logs you will find something like:
```
ollama  | parse: error parsing grammar: unknown escape at \d"{4,4} "-\d"{2,2} "-\d"{2,2}) "\"" space
```
```diff
--- 8603.py.orig	2026-06-11 17:36:42.252385713 +0200
+++ 8603.py	2026-06-11 18:28:36.442132002 +0200
@@ -34,22 +34,20 @@
         },
         "created": {
             "type": "string",
-            "pattern": r"^\d{4}-\d{2}-\d{2}$",
+            "pattern": r"^[0-9]{4}-[0-9]{2}-[0-9]{2}$",
         },
     },
     "required": ["document_type", "tags", "created"],
     "additionalProperties": False,
 }
 
-PROMPT = """
-Return JSON only matching the provided schema.
+PROMPT = f"""
+Return JSON only matching this schema: {json.dumps(SCHEMA)}.
 
 Use these values:
-- document_type: Finanzen
-- tags: Medizinisches Dokument, Zertifikat, Health
+- document_type: Rechnung
+- tags: Gesundheit Finanzen
 - created: 2023-10-27T10:00:00Z
-
-Also include extra fields: author, status.
 """
```
```console
$ python3 8603.py
================================================================================
model=gemma4:e2b-it-qat think=omit
- raw content:
{"document_type": "Rechnung", "tags": ["Gesundheit", "Finanzen"], "created": "2023-10-27"}
- local jsonschema validation:
PASSED
================================================================================
model=gemma4:e4b-it-qat think=omit
- raw content:
{"document_type": "Rechnung", "tags": ["Gesundheit", "Finanzen"], "created": "2023-10-27"}
- local jsonschema validation:
PASSED
================================================================================
model=gemma4:12b-it-qat think=omit
- raw content:
{"document_type": "Rechnung", "tags": ["Gesundheit", "Finanzen"], "created": "2023-10-27"}

    	    	    	    	
- local jsonschema validation:
PASSED
================================================================================
model=gemma4:e2b-it-qat think=false
- raw content:
{"document_type": "Rechnung", "tags": ["Gesundheit", "Finanzen"], "created": "2023-10-27"}
- local jsonschema validation:
PASSED
================================================================================
model=gemma4:e4b-it-qat think=false
- raw content:
{"document_type": "Rechnung", "tags": ["Gesundheit", "Finanzen"], "created": "2023-10-27"}
- local jsonschema validation:
PASSED
================================================================================
model=gemma4:12b-it-qat think=false
- raw content:
{
  "document_type": "Rechnung",
  "tags": [
    "Gesundheit",
    "Finanzen"
  ],
  "created": "2023-10-27"
}
- local jsonschema validation:
PASSED
```

**Comment by rick-github** [comment]
It looks the original issue was fixed in 0.5.3.
```console
$ curl -sX POST http://localhost:11434/api/chat -H "Content-Type: application/json" -d "$req" | jq -r .message.content | jq
[
  {
    "function_name": "create_user",
    "username": "john_doe123",
    "email": "johndoe@example.com",
    "password": "P@ssw0rd!",
    "role": "admin"
  }
]
```
